File: src/voltadomar/controller/controller.py
"""
controller/controller.py

Controller manages state for agent connections, session information,
and program execution. Handles coordination between users and agents.

Author: David Song <davsong@cs.washington.edu>
"""

import asyncio
import logging
from typing import Dict, Any as AnyType, Optional, Set

import grpc.aio
from google.protobuf.any_pb2 import Any

import voltadomar.anycast.anycast_pb2_grpc as anycast_pb2_grpc
from voltadomar.anycast.anycast_pb2 import (
    ReplyPayload,
    DonePayload,
    Message,
    ErrorPayload,
    RegisterPayload,
    Response,
    JobPayload,
)

from voltadomar.controller.program import Program
from voltadomar.controller.traceroute import (
    TracerouteConf,
    Traceroute,
    parse_traceroute_args,
)

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    handlers=[logging.StreamHandler()],
)
logger = logging.getLogger(__name__)


class Controller(anycast_pb2_grpc.AnycastServiceServicer):
    """
    Controller manages agent connections, session allocation, and
    program execution. Handles communication between users and agents.
    """

    def __init__(self, port: int, start_id: int, end_id: int, block_size: int):
        """
        Initialize a new Controller instance.

        This class manages traceroute sessions and agent streams, keeping track of session IDs
        and maintaining mappings between programs and their corresponding agents.

        Args:
            port (int): The port to run the server on
            start_id (int): The starting session ID number to use for new traceroute sessions
            end_id (int): The maximum session ID number that can be allocated
            block_size (int): The size of ID blocks to allocate for each agent
        """
        self.port = port
        self.start_id = start_id
        self.end_id = end_id
        self.block_size = block_size

        self.next_session_id = self.start_id
        self.programs: Dict[int, Program] = {}
        self.agent_streams: Dict[str, grpc.aio.ServicerContext] = {}
        self.in_use_blocks: Set[int] = set()

    def allocate_session_id(self) -> int:
        """
        Allocate a new session ID block using a sliding window approach.
        Returns the start ID of the allocated block.

        Raises:
            RuntimeError: If no free blocks are available.
        """
        current_id = self.next_session_id

        # Sweep through the entire ID range to find the next free block.
        for _ in range((self.end_id - self.start_id) // self.block_size):
            if current_id not in self.in_use_blocks:
                self.in_use_blocks.add(current_id)
                self.next_session_id = (current_id + self.block_size) % self.end_id
                if self.next_session_id < self.start_id:
                    # Handle wrap-around case.
                    self.next_session_id = self.start_id
                return current_id

            current_id = (current_id + self.block_size) % self.end_id
            if current_id < self.start_id:
                # Handle wrap-around case.
                current_id = self.start_id

        logger.error("No free session ID blocks available")
        # Error is propagated to the caller via try/catch.
        raise RuntimeError("No free session ID blocks available")

    def release_session_id(self, session_id: int) -> None:
        """
        Release a previously allocated session ID block.

        Args:
            session_id: The start ID of the block to release
        """
        if session_id in self.in_use_blocks:
            self.in_use_blocks.remove(session_id)
            logger.debug(f"Released session ID block starting at {session_id}")
        else:
            logger.warning(
                f"Attempted to release unallocated session ID block: {session_id}"
            )

    async def ControlStream(
        self, request_iterator: AnyType, context: grpc.aio.ServicerContext
    ) -> None:
        """
        Handle a stream of control messages from agents.

        Args:
            request_iterator: An async iterator of incoming messages from the agent.
            context: The gRPC context for the agent connection.
        """
        agent_id: Optional[str] = None
        try:
            async for message in request_iterator:
                match message.type:
                    case "REGISTER":
                        register_payload = RegisterPayload()
                        message.payload.Unpack(register_payload)
                        agent_id = register_payload.agent_id
                        logger.debug(f"REGISTER from {agent_id}")

                        self.agent_streams[agent_id] = context
                        logger.info(f"Agent {agent_id} registered")

                    case "DONE":
                        if agent_id is None:
                            logger.error("Received DONE without preceding REGISTER")
                            continue
                        done_payload = DonePayload()
                        message.payload.Unpack(done_payload)
                        session_id = done_payload.session_id
                        logger.debug(f"DONE from {agent_id} for session {session_id}")
                        if session_id in self.programs:
                            self.programs[session_id].handle_done(done_payload)
                        else:
                            logger.error(
                                f"Received DONE for unknown session ID {session_id}"
                            )

                    case "REPLY":
                        if agent_id is None:
                            logger.error("Received REPLY without preceding REGISTER")
                            continue
                        logger.debug(f"REPLY from {agent_id}")
                        reply_payload = ReplyPayload()
                        message.payload.Unpack(reply_payload)
                        # Unlike DONE, We don't know the session ID for REPLY messages, so we
                        # need to check all programs to see if any of them can handle the reply.
                        for program in self.programs.values():
                            if program.handle_reply(reply_payload, agent_id):
                                break
                        else:
                            logger.error("Received REPLY for unknown session ID")

                    case "ERROR":
                        error_payload = ErrorPayload()
                        message.payload.Unpack(error_payload)
                        logger.error(
                            f"Error from {agent_id} code: {error_payload.code} message: {error_payload.message}"
                        )

                    case _:
                        logger.error(f"Unknown message type: {message.type}")
                        continue

        except KeyboardInterrupt:
            logger.info("Keyboard interrupt received. Shutting down.")
            raise
        except grpc.aio.AioRpcError as e:
            logger.info(f"Agent {agent_id or 'unknown'} disconnected: {e}")
        except Exception as e:
            logger.exception(
                f"Error processing agent stream for {agent_id or 'unknown'}: {e}"
            )
        finally:
            self.agent_streams.pop(agent_id, None)

    async def UserRequest(
        self, request: AnyType, context: grpc.aio.ServicerContext
    ) -> Response:
        """
        Handle a user request to run a program (e.g., traceroute).

        Args:
            request: The user request containing the command to run.
            context: The gRPC context for the user connection.

        Returns:
            A Response object containing the result or an error message.
        """
        session_start_id: Optional[int] = None
        try:
            command = request.command
            args = parse_traceroute_args(command)
            source, destination, max_ttls, waittime, tos, nqueries = args
            if max_ttls * nqueries > self.block_size:
                return Response(
                    code=400,
                    output=f"Max hops * probe num ({max_ttls} * {nqueries}) exceeds block size ({self.block_size})",
                )
            session_start_id = self.allocate_session_id()

            conf = TracerouteConf(
                source_host=source,
                destination_host=destination,
                start_id=session_start_id,
                max_ttl=max_ttls,
                timeout=waittime,
                probe_num=nqueries,
                tos=tos,
            )

            self.programs[session_start_id] = Traceroute(self, conf)
            logger.info(f"Starting program {session_start_id} with command: {command}")
            output = await self.programs[session_start_id].run()
            return Response(code=200, output=output)
        except KeyboardInterrupt:
            logger.info("Keyboard interrupt received during user request.")
            raise
        except ValueError as e:
            logger.warning(f"Invalid user request: {e}")
            return Response(code=400, output=str(e))
        except grpc.aio.AioRpcError as e:
            logger.info(f"User disconnected: {e}")
        except Exception as e:
            logger.exception(f"Error processing user request: {e}")
            return Response(code=500, output=f"Internal server error: {e}")
        finally:
            self.programs.pop(session_start_id, None)
            self.release_session_id(session_start_id)

    async def send_job(self, agent_id: str, job_payload: JobPayload) -> None:
        """
        Send a job payload to a specific agent.

        Args:
            agent_id: The ID of the target agent.
            job_payload: The job payload protobuf message.
        """
        logger.debug(f"Sending job to agent {agent_id}")
        context = self.agent_streams.get(agent_id)
        if not context:
            logger.error(
                f"Attempted to send job to unknown or disconnected agent {agent_id}"
            )
            return

        packed_payload = Any()
        packed_payload.Pack(job_payload)
        message = Message(type="JOB", payload=packed_payload)

        try:
            await context.write(message)
            logger.info(f"Successfully sent job to agent {agent_id}")
        except grpc.aio.AioRpcError as e:
            logger.error(f"Failed to send job to agent {agent_id}: {e}")

    async def run(self) -> None:
        """
        Start the gRPC server and wait for termination.
        """
        server = grpc.aio.server()
        anycast_pb2_grpc.add_AnycastServiceServicer_to_server(self, server)
        server.add_insecure_port(f"0.0.0.0:{self.port}")
        await server.start()
        logger.info(f"Server started on port {self.port}")

        try:
            await server.wait_for_termination()
        except asyncio.CancelledError:
            logger.info("Controller task was cancelled.")
            raise
        finally:
            await server.stop(0)
            logger.info("Controller shutdown complete.")

File: src/voltadomar/internal/controller/controller.go
// Package controller implements the Voltadomar controller functionality
package controller

import (
	"context"
	"fmt"
	"log"
	"net"
	"sync"

	"google.golang.org/grpc"
	"google.golang.org/protobuf/types/known/anypb"

	"github.com/randofan/voltadomar/internal/constants"
	pb "github.com/randofan/voltadomar/proto/anycast"
)

// Controller manages agent connections, session allocation, and program execution
type Controller struct {
	pb.UnimplementedAnycastServiceServer

	port      int
	startID   int32
	endID     int32
	blockSize int32

	// Session management
	nextSessionID int32
	inUseBlocks   map[int32]bool

	// Agent and program management
	agentStreams map[string]pb.AnycastService_ControlStreamServer
	programs     map[int32]Program

	// Synchronization
	mu sync.RWMutex

	// gRPC server
	server *grpc.Server
}

// NewController creates a new Controller instance with input validation
func NewController(port int, startID, endID, blockSize int32) (*Controller, error) {
	// Validate input parameters
	if port <= 0 || port > 65535 {
		return nil, fmt.Errorf("invalid port %d: must be between 1 and 65535", port)
	}
	if startID < 0 {
		return nil, fmt.Errorf("invalid startID %d: must be non-negative", startID)
	}
	if endID <= startID {
		return nil, fmt.Errorf("invalid endID %d: must be greater than startID %d", endID, startID)
	}
	if blockSize <= 0 {
		return nil, fmt.Errorf("invalid blockSize %d: must be positive", blockSize)
	}
	if blockSize > (endID - startID) {
		return nil, fmt.Errorf("blockSize %d is larger than available range %d", blockSize, endID-startID)
	}

	return &Controller{
		port:          port,
		startID:       startID,
		endID:         endID,
		blockSize:     blockSize,
		nextSessionID: startID,
		inUseBlocks:   make(map[int32]bool),
		agentStreams:  make(map[string]pb.AnycastService_ControlStreamServer),
		programs:      make(map[int32]Program),
	}, nil
}

// AllocateSessionID allocates a new session ID block using a sliding window approach
func (c *Controller) AllocateSessionID() (int32, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Calculate total number of possible blocks
	totalRange := c.endID - c.startID
	maxBlocks := totalRange / c.blockSize
	if maxBlocks == 0 {
		return 0, fmt.Errorf("block size %d is larger than available range %d", c.blockSize, totalRange)
	}

	startSearch := c.nextSessionID
	currentID := startSearch
	blocksChecked := int32(0)

	// Search for available block with bounded loop
	for blocksChecked < maxBlocks {
		// Ensure currentID is within bounds and aligned to block boundaries
		if currentID < c.startID || currentID+c.blockSize > c.endID {
			currentID = c.startID
		}

		if !c.inUseBlocks[currentID] {
			c.inUseBlocks[currentID] = true
			c.nextSessionID = currentID + c.blockSize
			if c.nextSessionID >= c.endID {
				c.nextSessionID = c.startID
			}
			log.Printf("Allocated session ID block starting at %d", currentID)
			return currentID, nil
		}

		currentID += c.blockSize
		blocksChecked++
	}

	return 0, fmt.Errorf("no available session ID blocks (checked %d blocks)", blocksChecked)
}

// ReleaseSessionID releases a session ID block
func (c *Controller) ReleaseSessionID(sessionID int32) {
	c.mu.Lock()
	defer c.mu.Unlock()

	delete(c.inUseBlocks, sessionID)
	log.Printf("Released session ID block starting at %d", sessionID)
}

// SendJob sends a job payload to a specific agent
func (c *Controller) SendJob(agentID string, jobPayload *pb.Job) error {
	c.mu.RLock()
	stream, exists := c.agentStreams[agentID]
	c.mu.RUnlock()

	if !exists {
		return fmt.Errorf("agent %s not connected", agentID)
	}

	// Pack the job payload
	anyPayload, err := anypb.New(jobPayload)
	if err != nil {
		return fmt.Errorf("failed to pack job payload: %w", err)
	}

	message := &pb.Message{
		Payload: anyPayload,
	}

	if err := stream.Send(message); err != nil {
		return fmt.Errorf("failed to send job to agent %s: %w", agentID, err)
	}

	log.Printf("Successfully sent job to agent %s", agentID)
	return nil
}

// ControlStream handles bidirectional streaming with agents
func (c *Controller) ControlStream(stream pb.AnycastService_ControlStreamServer) error {
	var agentID string

	defer func() {
		if agentID != "" {
			c.mu.Lock()
			delete(c.agentStreams, agentID)
			c.mu.Unlock()
			log.Printf("Agent %s disconnected", agentID)
		}
	}()

	for {
		message, err := stream.Recv()
		if err != nil {
			log.Printf("Error receiving message from agent %s: %v", agentID, err)
			return err
		}

		if err := c.processAgentMessage(message, &agentID, stream); err != nil {
			log.Printf("Error processing message from agent %s: %v", agentID, err)
			return err
		}
	}
}

// processAgentMessage processes a single message from an agent
func (c *Controller) processAgentMessage(message *pb.Message, agentID *string, stream pb.AnycastService_ControlStreamServer) error {
	if message.Payload == nil {
		return fmt.Errorf("received message with nil payload")
	}

	switch message.Payload.TypeUrl {
	case constants.MsgTypeRegister:
		registerPayload := &pb.Register{}
		if err := message.Payload.UnmarshalTo(registerPayload); err != nil {
			return fmt.Errorf("failed to unmarshal Register: %w", err)
		}

		*agentID = registerPayload.AgentId
		c.mu.Lock()
		c.agentStreams[*agentID] = stream
		c.mu.Unlock()

		log.Printf("Agent %s registered", *agentID)

	case constants.MsgTypeDone:
		donePayload := &pb.Done{}
		if err := message.Payload.UnmarshalTo(donePayload); err != nil {
			return fmt.Errorf("failed to unmarshal Done: %w", err)
		}

		c.mu.RLock()
		program, exists := c.programs[donePayload.SessionId]
		c.mu.RUnlock()

		if exists {
			program.HandleDone(donePayload)
		}

	case constants.MsgTypeReply:
		replyPayload := &pb.Reply{}
		if err := message.Payload.UnmarshalTo(replyPayload); err != nil {
			return fmt.Errorf("failed to unmarshal Reply: %w", err)
		}

		// Try to find a program that can handle this reply
		c.mu.RLock()
		for _, program := range c.programs {
			if program.HandleReply(replyPayload, *agentID) {
				break
			}
		}
		c.mu.RUnlock()

	case constants.MsgTypeError:
		errorPayload := &pb.Error{}
		if err := message.Payload.UnmarshalTo(errorPayload); err != nil {
			return fmt.Errorf("failed to unmarshal Error: %w", err)
		}

		log.Printf("Error from agent %s: %d - %s", *agentID, errorPayload.Code, errorPayload.Message)

	default:
		messageType := constants.GetMessageTypeFromURL(message.Payload.TypeUrl)
		log.Printf("Unknown message type from agent %s: %s", *agentID, messageType)
	}

	return nil
}

// UserRequest handles user requests to run programs
func (c *Controller) UserRequest(ctx context.Context, request *pb.Request) (*pb.Response, error) {
	log.Printf("Received user request: %s", request.Command)

	// Allocate session ID
	sessionStartID, err := c.AllocateSessionID()
	if err != nil {
		return &pb.Response{
			Code:   500,
			Output: fmt.Sprintf("Failed to allocate session ID: %v", err),
		}, nil
	}

	defer c.ReleaseSessionID(sessionStartID)

	// Parse traceroute command
	conf, err := ParseTracerouteArgs(request.Command)
	if err != nil {
		return &pb.Response{
			Code:   400,
			Output: fmt.Sprintf("Invalid command: %v", err),
		}, nil
	}

	conf.StartID = sessionStartID

	// Create and run traceroute program
	traceroute, err := NewTraceroute(c, conf)
	if err != nil {
		return &pb.Response{
			Code:   500,
			Output: fmt.Sprintf("Failed to create traceroute: %v", err),
		}, nil
	}

	// Register the program
	c.mu.Lock()
	c.programs[sessionStartID] = traceroute
	c.mu.Unlock()

	defer func() {
		c.mu.Lock()
		delete(c.programs, sessionStartID)
		c.mu.Unlock()
	}()

	// Run the program
	output, err := traceroute.Run()
	if err != nil {
		return &pb.Response{
			Code:   500,
			Output: fmt.Sprintf("Program execution failed: %v", err),
		}, nil
	}

	return &pb.Response{
		Code:   200,
		Output: output,
	}, nil
}

// Run starts the gRPC server and waits for termination
func (c *Controller) Run() error {
	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", c.port))
	if err != nil {
		return fmt.Errorf("failed to listen on port %d: %w", c.port, err)
	}

	c.server = grpc.NewServer()
	pb.RegisterAnycastServiceServer(c.server, c)

	log.Printf("Controller starting on port %d", c.port)
	log.Printf("Session ID range: %d-%d, block size: %d", c.startID, c.endID, c.blockSize)

	if err := c.server.Serve(lis); err != nil {
		return fmt.Errorf("failed to serve: %w", err)
	}

	return nil
}

// Stop gracefully stops the controller
func (c *Controller) Stop() {
	if c.server != nil {
		log.Println("Stopping controller...")
		c.server.GracefulStop()
		log.Println("Controller stopped")
	}
}
